File: Java/BaekJoon/Gold/G3_2177.java
package BaekJoon.Gold;

import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

class COIN {
    long cost;
    long cnt;

    public COIN(long cost, long cnt) {
        this.cost = cost;
        this.cnt = cnt;
    }
}

public class G3_2177 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        long C = sc.nextLong();

        COIN[] coins = new COIN[N];

        for (int i = 0; i < N; i++) {
            long cost = sc.nextLong();
            long cnt = sc.nextLong();

            coins[i] = new COIN(cost, cnt);
        }

        Arrays.sort(coins, new Comparator<COIN>() {
            @Override
            public int compare(COIN left, COIN right) {
                return Long.compare(right.cost, left.cost);
            }
        });

        long ans = 0;

        while(true) {
            long pay = C;
            
            // 최저 급료 정확하게 지급
            for (int i = 0; i < N; i++) {
                if (pay < coins[i].cost) continue;

                long use = Math.min(coins[i].cnt, pay / coins[i].cost);

                pay = pay - coins[i].cost * use;
                coins[i].cnt -= use;
            }

            // 최저 급료 초과 지급
            if (pay > 0) {
                boolean overpay = false;
                for (int i = N - 1; i >= 0; i--) {
                    if (coins[i].cnt > 0) {
                        pay = pay - coins[i].cost;
                        coins[i].cnt -= 1;

                        overpay = true;
                        break;
                    }
                }

                if (!overpay) break;
            }

            ans++;
        }

        System.out.println(ans);

        sc.close();
    }
}

File: Cpp/BaekJoon/Gold/G3_2177.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Coins {
  long long cost;
  long long cnt;
};

bool comp(const Coins& left, const Coins& right) {
  if (left.cost != right.cost) return left.cost > right.cost;

  return false;
}

int main() {
  int N;
  long long C;

  cin >> N >> C;

  vector<Coins> coins;
  coins.resize(N);

  for (int i = 0; i < N; i++) {
    cin >> coins[i].cost >> coins[i].cnt;
  }

  sort(coins.begin(), coins.end(), comp);

  long long ans = 0;

  while (true) {
    long long pay = C;

    // 최저급료에 정확히 맞춰서 지급
    for (int i = 0; i < N; i++) {
      if (pay < coins[i].cost) continue;

      // 사용한 동전의 개수
      long long use = min(coins[i].cnt, pay / coins[i].cost);

      // 남은 급료 및 현금 개수
      pay = pay - (coins[i].cost * use);
      coins[i].cnt = coins[i].cnt - use;
    }

    // 정확히 지급하지 못하는 경우 초과 지급 계산
    if (pay > 0) {
      bool overpay = false;

      for (int i = N - 1; i >= 0; i--) {
        if (coins[i].cnt > 0) {
          coins[i].cnt = coins[i].cnt - 1;
          pay = pay - coins[i].cost;
          overpay = true;

          break;
        }
      }

      if (!overpay) break;
    }

    ans++;
  }

  cout << ans;

  return 0;
}